File: rcs/riot-sdk/python/riot_sdk/core/order_snapshots.py
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum


class OrderLookupStatus(str, Enum):
    """Semantic result of observing an order by caller-owned upper_id."""

    Found = "Found"
    NotFound = "NotFound"
    AbsentAtObservation = "AbsentAtObservation"
    Indeterminate = "Indeterminate"


@dataclass(frozen=True, slots=True)
class OrderMissionSnapshot:
    type: str | None
    map_id: int | None
    destination: int | None


@dataclass(frozen=True, slots=True)
class OrderSnapshot:
    id: int
    order_id: str
    upper_id: str
    order_state: int
    appoint_vehicle_key: str | None
    execute_vehicle_key: str | None
    end_station_no: int | None
    missions: tuple[OrderMissionSnapshot, ...]


@dataclass(frozen=True, slots=True)
class OrderLookupResult:
    requested_upper_id: str
    status: OrderLookupStatus
    order: OrderSnapshot | None = None

    def __post_init__(self) -> None:
        if not self.requested_upper_id or not self.requested_upper_id.strip():
            raise ValueError("requested_upper_id is required")
        if self.status is OrderLookupStatus.Found:
            if self.order is None:
                raise ValueError("Found requires an order")
            if self.order.upper_id != self.requested_upper_id:
                raise ValueError("Found order upper_id must match requested_upper_id")
        elif self.order is not None:
            raise ValueError(f"{self.status.value} must not carry an order")

    @classmethod
    def found(cls, requested_upper_id: str, order: OrderSnapshot) -> OrderLookupResult:
        return cls(requested_upper_id, OrderLookupStatus.Found, order)

    @classmethod
    def not_found(cls, requested_upper_id: str) -> OrderLookupResult:
        return cls(requested_upper_id, OrderLookupStatus.NotFound)

    @classmethod
    def absent_at_observation(cls, requested_upper_id: str) -> OrderLookupResult:
        return cls(requested_upper_id, OrderLookupStatus.AbsentAtObservation)

    @classmethod
    def indeterminate(cls, requested_upper_id: str) -> OrderLookupResult:
        return cls(requested_upper_id, OrderLookupStatus.Indeterminate)

File: rcs/riot-sdk/csharp/RIoT.Sdk.Core/OrderSnapshots.cs
namespace RIoT.Sdk.Core;

/// <summary>
/// Semantic result of observing one order by its caller-owned upperId.
/// </summary>
public enum OrderLookupStatus
{
    Found,
    NotFound,
    AbsentAtObservation,
    Indeterminate,
}

/// <summary>
/// Stable mission facts projected from a RIoT order response.
/// </summary>
public sealed record OrderMissionSnapshot(
    string? Type,
    int? MapId,
    int? Destination);

/// <summary>
/// Stable order facts exposed without leaking Kiota-generated types.
/// </summary>
public sealed record OrderSnapshot(
    long Id,
    string OrderId,
    string UpperId,
    int OrderState,
    string? AppointVehicleKey,
    string? ExecuteVehicleKey,
    int? EndStationNo,
    IReadOnlyList<OrderMissionSnapshot> Missions);

/// <summary>
/// Lookup result that keeps an observed absence distinct from authoritative not-found.
/// </summary>
public sealed record OrderLookupResult
{
    private OrderLookupResult(
        string requestedUpperId,
        OrderLookupStatus status,
        OrderSnapshot? order)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(requestedUpperId);
        RequestedUpperId = requestedUpperId;
        Status = status;
        Order = order;
    }

    public string RequestedUpperId { get; }
    public OrderLookupStatus Status { get; }
    public OrderSnapshot? Order { get; }

    public static OrderLookupResult Found(string requestedUpperId, OrderSnapshot order)
    {
        ArgumentNullException.ThrowIfNull(order);
        if (!string.Equals(requestedUpperId, order.UpperId, StringComparison.Ordinal))
        {
            throw new ArgumentException(
                "Found order upperId must match requestedUpperId.",
                nameof(order));
        }
        return new(requestedUpperId, OrderLookupStatus.Found, order);
    }

    public static OrderLookupResult NotFound(string requestedUpperId) =>
        new(requestedUpperId, OrderLookupStatus.NotFound, null);

    public static OrderLookupResult AbsentAtObservation(string requestedUpperId) =>
        new(requestedUpperId, OrderLookupStatus.AbsentAtObservation, null);

    public static OrderLookupResult Indeterminate(string requestedUpperId) =>
        new(requestedUpperId, OrderLookupStatus.Indeterminate, null);
}
